### PR TITLE
push: Prepare pack before sending pack header.

### PR DESCRIPTION
**Pull request by ccstolley**
For large pushes, preparing the pack can take a while. Currently we send the pack header first, followed by preparing the pack and then finally sending the pack. Unfortunately github.com will terminate a git-receive-pack command over http if it is idle for more than 10 seconds. This is easily exceeded for a large push, and so the push is rejected with a Broken Pipe error.

This patch moves the pack preparation ahead of sending the pack header, so that the timeout is avoided.

`prepare_pack()` can be called multiple times but will only do the work once, so the original `PREPARE_PACK` call inside `git_packbuilder_foreach()` remains.

Resolves https://github.com/libgit2/libgit2/issues/5905 .

### CODE REVIEW

**Comment by ccstolley** [comment]
> it shows that there is a deficiency in the public API of the packbuilder

Are there specific use cases you had in mind for such an API? I'm reluctant to introduce a new API without a concrete justification for it. Perhaps I'm thinking too narrowly, but to me preparing the pack is a lower level detail that should be abstracted away by the library. And this bug illustrates that getting that timing right can be subtle.

File: src/pack-objects.h [hunk 1]
```diff
@@ -97,5 +97,7 @@ struct git_packbuilder {
 };
 
 int git_packbuilder__write_buf(git_str *buf, git_packbuilder *pb);
+int prepare_pack(git_packbuilder *pb);
```
Comment (ethomson) [inline src/pack-objects.h:100]: I don't think that we need to expose this publicly (yet) - I agree that exposing APIs that aren't "needed" gives us a maintenance burden moving forward.  But let's give it a nicer name even if it's an internal API.

```suggestion
int git_packbuilder__prepare(git_packbuilder *pb);
```

**Comment by ethomson** [comment]
LGTM.  I don't think that we need to make it public yet, necessarily - but I do think that we should expect that we might want to in the future if we have more data.  In the meantime, let's tighten up the internal name a bit and 🚢

**Comment by ethomson** [comment]
Thanks @ccstolley - looking good!  😎

**Pull request merged**